Restructure the producer doc (#589)
apache/pulsar-site: docs/client-libraries-producers.md

</Tabs>
````

## Publish messages

> This example shows how to publish messages using producers. The publish operation will be done until the broker tells us that the message has been successfully published. It will return the message id after the message is published successfully.

Reviewer: Suggest moving this introductory paragraph next to the code example.

Reviewer 2: +1

> This example shows how to publish messages using producers. The publish operation will be done until the broker tells us that the message has been successfully published. It will return the message id after the message is published successfully.

Reviewer 2: ```suggestion
This example shows how to publish messages using producers. The publish operation is done until the broker tells you the message has been successfully published. Pulsar returns the message ID after the message is published successfully.
```

1. Who returns the msg ID? Pulsar? Need to specify
2. Writing style: 
a. Write in the simple present tense as much as possible if you are covering facts that were, are, and forever shall be true. https://docs.google.com/document/d/1lc5j4RtuLIzlEYCBo97AC8-U_3Erzs_lxpkDuseU0n4/edit#bookmark=id.e8uqh1awkcnp
b. Avoid the first-person pronouns I and we https://docs.google.com/document/d/1lc5j4RtuLIzlEYCBo97AC8-U_3Erzs_lxpkDuseU0n4/edit#bookmark=id.hyip4lzbapvm

Author: Thanks for your guidance. 

> Who returns the msg ID? Pulsar? Need to specify

It's the broker. I have fixed it.

> Pulsar supports both synchronous and asynchronous publishing of messages in most language clients. In some languages client(`Node.js` and `C#`), it is easy to publish messages synchronously based on the asynchronous method using language-specific mechanisms(like `await`).

Reviewer: ```suggestion
Pulsar supports both synchronous and asynchronous publishing of messages in most clients. In some language-specific clients, such as Node.js and C#, you can publish messages synchronously based on the asynchronous method using language-specific mechanisms (like `await`).
```

   ```go
    msg := pulsar.ProducerMessage{
        Payload: []byte("my-sync-message"),
    }

    if _, err := producer.send(msg); err != nil {
      log.Fatalf("Could not publish message due to: %v", err)
    }
   ```

>   For all methods of the `ProducerMessage` object, see [here](https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar#ProducerMessage).

Reviewer 2: ```suggestion
  For all methods of the `ProducerMessage` object, see [Go API doc](https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar#ProducerMessage).
```
Show the doc name explicitly instead of "here" for screen readers' convenience.

</Tabs>
````

## Configure messages

> There are various properties of Pulsar's messages that we can set. The values of these properties will be stored in the metadata of the message.

Reviewer: ```suggestion
You can set various properties of Pulsar's messages. The values of these properties are stored in the metadata of a message.
```

````mdx-code-block
<Tabs groupId="lang-choice"
  defaultValue="Java"
  values={[{"label":"Java","value":"Java"},{"label":"C++","value":"C++"},{"label":"C#","value":"C#"}]}>

<TabItem value="Java">

  ```java
  producer.newMessage()
                .key("my-key") // Set the message key
                .eventTime(System.currentTimeMillis()) // Set the event time
                .sequenceId(1203) // Set the sequenceId for the deduplication purposes
                .deliverAfter(1, TimeUnit.HOURS) // Delay message delivery for 1 hour
                .property("my-key", "my-value") // Set the customized metadata
                .property("my-other-key", "my-other-value")
                .replicationClusters(
                        Lists.newArrayList("r1", "r2")) // Set the geo-replication clusters for this message.
                .value("content")
                .send();
  ```

>   For the Java client, we can also use `loadConf` to configure the message metadata. Here is the example:

Reviewer: ```suggestion
  For the Java client, you can also use `loadConf` to configure the message metadata. Here is an example:
```

public interface MessageRouter extends Serializable {
    int choosePartition(Message msg);
}
```

The following router routes every message to partition 10:

```java
public class AlwaysTenRouter implements MessageRouter {
    public int choosePartition(Message msg) {
        return 10;
    }
}
```

With that implementation, you can send messages to partitioned topics as below.

```java
Producer<byte[]> producer = pulsarClient.newProducer()
        .topic("my-topic")
        .messageRouter(new AlwaysTenRouter())
        .create();
producer.send("Partitioned topic message".getBytes());
```

  </TabItem>

  <TabItem value="C++">

> To use a custom message router, you need to provide an implementation of the `MessageRoutingPolicy· interface, which has just one `getPartition` method:

Reviewer: ```suggestion
To use a custom message router, you need to provide an implementation of the `MessageRoutingPolicy` interface, which has one `getPartition` method:
```

};
```

With that implementation, you can send messages to partitioned topics as below.
```cpp
Producer producer;
Result result = client.createProducer(
    "my-topic",
    ProducerConfiguration().setMessageRouter(std::make_shared<MessageRouter>()),
    producer);
Message msg = MessageBuilder().setContent("content").build();
result = producer.send(msg);
```
  </TabItem>

  <TabItem value="Go">

>   In the Go client, we can configure the customize message router by passing a funciton.

Reviewer: ```suggestion
  In the Go client, you can configure a customized message router by passing a function.
```
